### ISSUE TITLE
incorrect intersection results from same two objects in different relative positions

### ISSUE THREAD

**Issue by steveliles** [samalexander/three-csg-ts#21]
When intersecting the same two objects with different relative positions, certain positions result in an incorrect intersection object.

**How to reproduce** 
At this point it is not clear to me what is a minimal repro. I am intersecting a pyramidal shape representing a camera frustum with an L shaped "room" object. 

When the camera is at certain positions in the room the returned intersection is clearly incorrect while the individual room and frustum objects (when rendered separately) are correct. 

Possibly I am using CSG incorrectly, but then I cannot explain why other positions in the room do render correct intersections.

Both of the objects I'm intersecting are created programmatically. The room is an ExtrudeBufferGeometry extruded from a programmatically created THREE.Shape, while the frustum is a BufferGeometry with vertices set from a dynamically created array like this: 

```
  const geometry = new THREE.BufferGeometry()
  geometry.setAttribute('position', new THREE.BufferAttribute(vertices, 3))
  geometry.computeVertexNormals()
```

**Screenshots**
animated gif showing that in some positions everything works great, while in others there's a clear problem:
![fov](https://user-images.githubusercontent.com/441745/114313255-b2a4bf00-9aed-11eb-9c23-1986c891443a.gif)

In positions that cause the intersection to break, the individual objects render as expected:

1. frustum only (green):
![Screenshot from 2021-04-11 17-04-56](https://user-images.githubusercontent.com/441745/114313335-0a432a80-9aee-11eb-84d0-69b8f4fab402.png)

2. lines-of-sight / room object only (red):
![Screenshot from 2021-04-11 17-05-56](https://user-images.githubusercontent.com/441745/114313356-19c27380-9aee-11eb-8db7-cb7538858c82.png)

wireframe materials illustrating the broken intersection: room (red), frustum (yellow), intersection (green)
![Screenshot from 2021-04-11 17-33-54](https://user-images.githubusercontent.com/441745/114313405-4d9d9900-9aee-11eb-97cb-e0a208d67ee1.png)

**Additional context**
I'm using react-three-fiber, but I don't think this is relevant as the CSG code is outside of react and dealing directly with threejs objects:

```
export const createConstrainedFOVMesh = props => {
  const { id } = props
  const fovMesh = createSimpleFOVMesh(props)
  const linesOfSightMesh = createLinesOfSightMesh(props)

  if (linesOfSightMesh) {
    const bspLinesOfSight = CSG.fromMesh(linesOfSightMesh)
    const bspFOV = CSG.fromMesh(fovMesh)

    const bspResult = bspFOV.intersect(bspLinesOfSight)
    const result = CSG.toMesh(bspResult, fovMesh.matrix)
    result.material = fovMesh.material
    result.name = id

    return result
  }
  return fovMesh
}
```

**Comment by steveliles** [comment]
update: 

I oversimplified initially when I talked about the "room" object. It isn't really the shape of the room, it is a generated object that is a sub-set of the room that would be visible to a camera if it were to pan around (ie the "visibility polygon" or total visible portion of the room from the camera's position).

When the camera moves I have to update this object to reflect the visibility polygon from the camera's new position. This naturally creates different extruded geometries at different camera positions, and it appears that certain of those geometries are what cause the broken intersections, despite the extruded visibility polygon / room object being perfectly valid.

correct intersection:
![Screenshot from 2021-04-11 18-08-27](https://user-images.githubusercontent.com/441745/114314482-cbfc3a00-9af2-11eb-80bd-825c79474a76.png)

broken intersection (note that the faces of the "room" shape are different than previous image):
![Screenshot from 2021-04-11 18-08-13](https://user-images.githubusercontent.com/441745/114314494-dddddd00-9af2-11eb-9179-936b0ad8d6f7.png)

In all cases the individual objects for my visibility polygon and camera frustum render correctly. 

I suspect I could now create a minimal repro test-case if that is useful?

**Comment by steveliles** [comment]
replicated in codesandbox [here](https://codesandbox.io/s/three-csg-ts-issue-21-nyykd?file=/src/App.js) with a simple geometry extruded from a THREE.shape intersected with a THREE.ConeBufferGeometry.

![three-csg-issue-21](https://user-images.githubusercontent.com/441745/115160944-c1154c80-a092-11eb-8396-8f552b2b276a.gif)

**Comment by steveliles** [comment]
update: 

I tried several other three csg libs and encountered the same issue in all of them.

@darkwave found a neat workaround to the issue I reported in [THREE-CSGMesh](https://github.com/manthrax/THREE-CSGMesh/issues/10) by using a CylinderGeometry with a very small radiusTop. 

I have tested and confirm that this workaround also applies to three-csg-ts! ([sandbox](https://codesandbox.io/s/three-csg-ts-issue-21-forked-czn4y?file=/src/App.js))

@darkwave also noted that setting the cylinder's radiusTop to 0 exhibits the same problem as using a cone directly, which may shed some light on the source of the problem, and again this is true with three-csg-ts.

**Comment by samalexander** [comment]
@steveliles please could you confirm whether the issue is still present with v3.0.0?

**Comment by steveliles** [comment]
@samalexander confirmed - still present with 3.x (codesandbox showing problem with v3.1.3 [here](https://codesandbox.io/s/three-csg-ts-issue-21-nyykd)).

**Comment by manthrax** [comment]
Pretty sure you've found the cause. I haven't verified it directly, but it sounds like the ConeGeometry actually makes a zero sized end cap. That endcap has an undefined face normal, and probably blows the whole algorithm, because if multiple planes share edges with a degenerate faces, the whole recursive slicing operation would break since the object isn't considered a watertight solid any more.

We could perhaps write a routine to check for bad inputs.. but it's not something you would want to run all the time unless you can't control the shape of your data.

Such a routine could perhaps collapse degenerate faces into a single vertex..

**Comment by manthrax** [comment]
I'm gonna leave this open for now if someone wants to discuss it further? Thanks @steveliles and @darkwave for chasing this down... :)

**Comment by manthrax** [comment]
doing this cleanup on the bsp tree seems to fix this case for conegeometry:

`
    bspFrustum.polygons = bspFrustum.polygons.filter( p => !isNaN(p.plane.normal.x)  );
`

we could perhaps add this to CSG.fromMesh...

**Comment by manthrax** [comment]
I made a branch and pr with that filter/check added to CSG.fromMesh, but I haven't tested it very well..
I could use some eyeballs/testing if anyone has some cycles... @steveliles @samalexander @darkwave
https://github.com/Jiro-Digital/three-csg-ts/pull/30

Open to feedback/suggestions/opinions...

**Comment by samalexander** [comment]
@steveliles your test case appears to be fixed by v3.1.4 (credit to @manthrax for the fix) - see [updated sandbox](https://codesandbox.io/s/three-csg-ts-issue-21-forked-1umws). 

Please could you confirm this issue can now be closed?

**Comment by manthrax** [comment]
Let's close it! :D
